src/Microsoft.ML.Data/Data/Conversion.cs

```diff
@@ -1305,11 +1277,11 @@ private bool TryParseCore(string text, int ich, int lim, out ulong dst)
         /// </summary>
         public bool TryParse(ref TX src, out I1 dst)
         {
-            long res;
-            bool f = TryParseSigned(RawI1.MaxValue, ref src, out res);
-            Contracts.Assert(f || res == I1.RawNA);
-            Contracts.Assert((RawI1)res == res);
-            dst = (RawI1)res;
+            long? res;
+            bool f = TryParseSigned(I1.MaxValue, ref src, out res);
+            Contracts.Check(f && res.HasValue);
+            Contracts.Assert((I1)res == res);
+            dst = (I1)res;
```

Comment:
>dst = (I1)res; [](start = 12, length = 14)

Previously, if someone attempted to parse a value like "500" as I1, the result was `DvInt1.NA`. Under your change, you are first parsing that into a `long?` as 500, you will of course fail the first assert (but not from a user's perspective since it will be in debug builds only), and from a user's perspective it will happily store the value 244.

Anyway I think you need to think a bit harder about this code, as well as introduce some tests from these text parsing scenarios. I think we have some in the private repo, maybe you should migrate them since you're changing this code a lot.

Reply:
yeah we need to change that assert to check so that it fails from user's perspective. 

---
In reply to: [212040729](https://github.com/dotnet/machinelearning/pull/683#discussion_r212040729) [](ancestors = 212040729)